core/architecture.md

```diff
@@ -0,0 +1,150 @@
+updatable at runtime by simply updating the policy. This different from NSM
+for three reasons:
+
+1. Nimbess only ever uses a single high-speed port for a pod
+2. UNP does not try to do VNF/CNF orchestration with K8S
+3. UNP is not restricted to L2-L3 service connectivity/policy
+
+In container networking, there is no real reason to use more than one port to a
+container. Technically speaking, a port just points to a socket referencing
+virtual or physical memory address space. It does not matter if there is 1 or
+more ports between a data plane and a container. Functionally, it matters for
+traffic differentiation between a pod control plane, and perhaps one or more
+data plane networks it is connected to. However, the goal of Nimbess is to 
+simplify networking, and one way to do that is to do real traffic
+differentiation within the data plane itself. Therefore with UNP, one can set
+policy which ensures that bandwidth is reserved for a containers control or
+management plane, while also applying QoS for dataplane connectivity and
+slicing traffic correctly for multi-tenancy.
+
+UNP can implement a Service Function Chain of CNFs, but it does not manage
+CNFs or define them. UNP is restricted to what it does best, networking, and
+therefore does not dictate CNF definitions or try to orchestrate them. The
+process to create a chain would be for an operator to create the CNFs as
+regular pods within Kubernetes, and then declare a UNP that defines a service
+chain for those pods. Remember, the chain may be updated at anytime and include
+new or different SFC classification or CNFs, a key advantage of Nimbess.
+
+UNP is able to classify traffic from L2-L7. The goal of UNP is to offer service
+policy mesh as used with other applications like [Istio](https://istio.io)
+without the need for an Envoy sidecar proxy application. The goal is to
+consolidate this type of functionality in the high speed dataplane, which will
+give better performance and a single fabric for all network data traffic.
+
+## Networking in Nimbess
+
+Most CNIs today offer L3 CNI solutions where some form of routing is used to
+control traffic between pods, especially across nodes. Some of these CNIs use
+static routing, their own routing implementation, or well known routing
+protocols such as BGP, which offers high scalability, fast route convergence
+times, as well as features like Equal-Cost Multipath (ECMP), Anycast, and
+VPN support. Therefore for L3 networking, Nimbess leverages iBGP as the control
+plane protocol for routing traffic between pods. Alternatively, Nimbess also 
+supports full L2 fabric across the cluster. This is important as many CNFs
+are L2 and may require a full L2 fabric to function correctly.
+
+## Architecture
+
+Nimbess consists of several components ran as containers across Kubernetes:
+
+[![Nimbess Architecture](../img/Nimbess.png)](../img/Nimbess.png)
+
+### Nimbess ETCD
```

Comment:
Does this mean that Nimbess declarative state will run on a separate etcd cluster than the K8s one? If so, what's the reason for doing that?

Reply:
Yes. The etcd in k8s is used for storing infra information. we should be deploying our own data store to avoid issues like collision, security (permission) and availability.

Reply:
I see Nimbess Stargaze as a K8s controller watching the UNP, and then, you could store the extra information needed for the implementation of the network intent in a separate etcd cluster specific for Nimbess. Copying the same info that is stored in K8s etcd to a separate one doesn't look very eficient to me. Just opening this for debate, and I know Calico and Contiv do that.

Reply:
Your point is well taken that we should not be duplicating objects in databases if we don't need to. Our thought is that we may need to update the resources with specific Nimbess fields like the ID of the resource that we translate it to. For example, if we see a k8s service, and we translate that into a load balancer Nimbess network module, then we may want to update that k8s service resource with the ID of our load balancer so we can back reference it later. We haven't worked through the specifics of how that part will work yet, but will include you when we come up with the design.